Machine learning (#51)
mlsa-book/MLSA: book/machinelearning.qmd

::: {#cnj-lm}
Let $f_R : \calX \rightarrow \calY$ be the regression task of interest with $\calX \subseteq \Reals^n$ and $\calY \subseteq \Reals^n$.
Let $(\xx, \yy)$ be data such that $\xx \in \calX$ and $\yy \in \calY$.

Say the learner of interest is a linear regression model with learning algorithm:

$$
(\hat{\beta_0}, \hat{\beta_1}) := \argmin_{\beta_0,\beta_1} \Big\{\sum^n_{i=1} (y_i - \gamma\beta_0 - \beta_1 x_i)^2\Big\}\text{ for }i = 1,...,n
$$

and prediction algorithm:

$$
\hatg(x) = \gamma\hat{\beta_0} + \hat{\beta_1}x
$$

The learner hyperparameters are $\Theta = (\gamma)$ which can take values $0$ or $1$ and the parameters are $\theta = (\beta_0, \beta_1)$.
The learner is fit by passing $(\xx, \yy)$ to the learning algorithm and thus estimating $\hat{\theta}$ and $\hatg$.
A prediction, $\haty$, is made for a new observation by passing $x^* \in \calX$ to the fitted model $\hatg$.

:::

::::

## Evaluating and benchmarking {#sec-ml-eval}

To understand if a model is 'good', its predictions are evaluated with a *loss function*.
Loss functions assign a score to the discrepancy between predictions and true values, $L: \calY \times \calY \rightarrow \ExtReals$.

> Given real-world data, $(\XX^*, \yy^*)$, and a trained model $\hatf$, then the loss is given by $L(\hatf{\XX^*}, \yy^*) = L(\hatyy, \yy^*)$.

Reviewer: ```suggestion
Given (unseen) real-world data, $(\XX^*, \yy^*)$, and a trained model, $\hatf$, the loss is given by $L(\hatf(\XX^*), \yy^*) = L(\hatyy, \yy^*)$.
```

For a model to be useful, it should perform well in general, and not just for the data used for training and development, which is known as a model's *generalization error*.

> A model should not be deployed, i.e., manually or automatically used to make predictions, its generalization error must be estimated to understand if the model is 'good'.

Reviewer: ```suggestion
A model should not be deployed, that is, manually or automatically used to make predictions, unless its generalization error was estimated to be acceptable for a given context.
```

Estimation of the generalization error requires *data splitting*, which is the process of splitting available data, $\calD$, into *training data*, $\dtrain \subset \calD$, and *testing data*, $\dtest = \calD \setminus \dtrain$.
<!-- FIXME: ABOVE COULD BE BETTER WORDED -->

The simplest method to estimate the generalization error is to use *holdout resampling*, which is the process of partitioning the data into one training dataset and one testing dataset, with the model trained on the former and predictions made for the latter.
Using 2/3 of the data for training and 1/3 for testing is a common splitting ratio (@Kohavi1995).
When splitting, it is essential the data is partitioned randomly to ensure any information encoded in data ordering is removed.
Ordering is often important in the real-world, for example in healthcare data when patients are recorded in order of enrolment to a study.
Whilst ordering can provide useful information, it does not generalize to new, unseen data.
For example, the number of days a patient has been in hospital is more useful than the patient's index in the dataset, as the former could be calculated for a new patient whereas the latter is meaningless.

> Another example is in the `rats` dataset that was briefly described in @sec-surv.

Reviewer: ```suggestion
Another example is in the `rats` dataset that was briefly described in @sec-surv.
```

The `rats` data explores how a novel drug effects tumor incidence.
However, the data is ordered by rat litters with every three rats being in the same litter.
Hence if rats were to be bred or raised differently over time,  even if the `litter` column were removed, this information would still be encoded in the order of the dataset and could impact upon any findings.

> Randomly splitting breaks any possible association between order and outcome.

Reviewer: There are situations where random splitting is not good: Time-series, hierarchical data (where rats litter could be an example for), spatially correlated data, etc.. I think i would write that in iid case random sampling is good, in other contexts other strategies might be helpful -> cite our overview paper http://arxiv.org/abs/2310.15108.

<!-- FIXME: ABOVE IS ALSO BAD! -->

> Holdout resampling is a quick method to estimate the generalization error, however it is an imperfect method as, even with random splitting, it can still result in 'unlucky' splits.

Reviewer: Use hold-out when data is huge

Returning to the `rats` data, consider a split that results in only male rats in the training data, the model would then be unable to handle female rats in the testing data, which it had never encountered before.
Alternatively, consider a random split with an even sex balance, as only two out of the 150 rats have a tumor, there is a high probability that the training data will not include any male rat with a tumor.

> Hence, in the case of a male rat, the model is likely to ignore the treatment variable and instead just use sex as the most important predictor of tumor incidence.

Reviewer: The reason to use CV is because hold-out has high variance for low-dim data. The problems mentioned here could be solfed by stratified resampling

This is important, and often recommended, in survival analysis to ensure that the proportion of censoring in each fold is representative of the full dataset.
<!-- FIXME: CITE OUR SURVIVAL BENCHMARK PAPER -->

Repeating resampling experiments with multiple models is referred to as a *benchmark experiment*.
A benchmark experiment compares models by evaluating their performance on *identical* data, which means the same resampling strategy and folds should be used for all models.
Determining if one model is actually better than another is a surprisingly complex topic [@Demsar2006; @Dietterich1998; @Nadeau2003] and is out of scope for this book, instead any benchmark experiments performed in this book are purely for illustrative reasons and no results are expected to generalize outside of these experiments.

## Optimization {#sec-ml-opt}

@sec-ml-models introduced model hyperparameters, which control how training and prediction algorithms are run.
Setting hyperparameters is a critical part of model fitting and can significantly change model performance.
*Tuning* is the process of using internal benchmark experiment to automatically select the optimal hyper-parameter configuration.

> For example, the number of trees, $m_r$ in a random forest (@sec-surv-ml-models-ranfor) is a potential hyperparameter to tune.

Reviewer: ```suggestion
For example, the number of trees, $m_r$, in a random forest (@sec-surv-ml-models-ranfor) is a potential hyperparameter to tune.
```

Reviewer: my least favorite example for hyperparam in RF :-) as theory says it should be as high as necessary, but higher in general better

This hyperparameter may be tuned over a range of values, say $[10, 1000]$ or over a discrete subset, say $\{10, 100, 1000\}$, for now assume the latter.
Three random forests with $10$, $100$, and $1000$ trees respectively are compared in a benchmark experiment.
The number of trees that results in the model with the optimal performance is then selected for the hyperparameter value going forward.
*Nested resampling* is a common method to prevent overfitting that could occur from using overlapping data for tuning, training, or testing.
Nested resampling is the process of resampling the training set again for tuning and then the optimal model is refit on the entire training data (@fig-ml-nested).

![An illustration of nested resampling. The large blocks represent three-fold CV for the outer resampling for model evaluation and the small blocks represent four-fold CV for the inner resampling for HPO. The light blue blocks are the training sets and the dark blue blocks are the test sets. Image and caption from @Becker2024 (CC BY-NC-SA 4.0).](https://mlr3book.mlr-org.com/chapters/chapter4/Figures/mlr3book_figures-11.svg){#fig-ml-nested}

> ## Conclusion

Reviewer: both, key takeaways and further reading are a bit thin, but can be extended later

:::: {.callout-warning icon=false}

## Key takeaways

* Machine learning tasks define the predictive problem of interest;

> * Regression tasks make continuous predictions and classification tasks make discrete predictions;

Reviewer: not true for classif task that predicts probability, its bounded but not discrete
